### PR TITLE
Implement preventing auto-scroll on new output

### LINKED ISSUE #980: Feature Request: Pause output or scrolling on click (and make it a setting)
See also #408

### LINKED ISSUE #3863: Feature Request: Preserve scroll position while not "at bottom"
# Description of the new feature/enhancement

I tried searching for this but couldn't find any dupes, but please let me know if this is already possible!

It'd be great to (at least have an option to) preserve the terminal scroll position while you aren't scrolled to the bottom.

Eg. If I'm doing a build or something, which is logging out info every second, but then want to scroll up and take a look at an error message I noticed higher up, my view shouldn't be affected by new lines being printed to the terminal. Whereas currently the view jerks around as new stuff is appended to it, making it impossible to actually read any history.

If I then scroll back down "as far as you can go", that should then be a signal to "stick" the scroll with the latest output again.

Could also be nice to have a small button appear somewhere if new lines have been printed while you were scrolled up, which you could click to jump back to the "stuck to the bottom" mode.

Thanks!

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
Updates the Terminal's scroll response to new output. The Terminal will not automatically scroll if...
- a selection is active, or
- the viewport is at the bottom of the scroll history

## References
#2529 - Spec
#3863 - Implementation

## PR Checklist
* [X] Closes #980
* [X] Closes #3863
* [ ] Tests added/passed
* [ ] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments
Updates the `_scrollOffset` value properly in TerminalCore when the cursor moves. We calculate a new `_scrollOffset` based on if we are circling the buffer and how far below the mutable bottom is.

We specifically check for if a selection is active and if the viewport is at the bottom, then use that as a condition for deciding if we should update `_scrollOffset` to the new calculated value or 0 (the bottom of the scroll history).

## Validation Steps Performed
Manual testing. Though I should add automated tests.
- [X] new output
- [X] new output when circling
- [X] new output when circling and viewport is at the top

### CODE REVIEW

File: src/cascadia/TerminalApp/Profile.cpp [hunk 1]
```diff
@@ -794,6 +809,77 @@ std::string_view Profile::_SerializeCloseOnExitMode(const CloseOnExitMode closeO
     }
 }
 
+int32_t Profile::ParseSnapOnOutput(const Json::Value& json)
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Profile.cpp:812]: I need to add a doc.

Also, the implementation for this section is up for debate in the spec.

File: src/cascadia/TerminalApp/Profile.cpp [hunk 2]
```diff
@@ -794,6 +809,77 @@ std::string_view Profile::_SerializeCloseOnExitMode(const CloseOnExitMode closeO
...
+    }
+    else if (snapOnOutput == SnapOnOutputFlag::AtBottom)
+    {
+        return SnapOnOutputAtBottom;
+    }
+    else if (snapOnOutput == SnapOnOutputFlag::NoSelection)
+    {
+        return SnapOnOutputNoSelection;
+    }
+    return "";
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Profile.cpp:880]: We don't actually use the "serialize" functions anywhere (I think). So, I think this is incorrect, but I also don't know what the "correct" behavior should be. We're also not using it so I guess I'll just worry about this when we start serializing again?

File: src/inc/DefaultSettings.h [hunk 3]
```diff
@@ -42,4 +42,5 @@ constexpr COLORREF DEFAULT_CURSOR_COLOR = COLOR_WHITE;
 constexpr COLORREF DEFAULT_CURSOR_HEIGHT = 25;
 
 const std::wstring DEFAULT_WORD_DELIMITERS{ L" ./\\()\"'-:,.;<>~!@#$%^&*|+=[]{}~?\u2502" };
+constexpr int32_t DEFAULT_SNAP_ON_OUTPUT = 0x1 | 0x2;
```
Comment (carlos-zamora) [inline src/inc/DefaultSettings.h:45]: I don't like this but I don't know how to fix it. Or is it ok if I just remove it?

Comment (fdciabdul) [reply src/inc/DefaultSettings.h:45]: just remove it i thought

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 4]
```diff
@@ -791,11 +803,41 @@ void Terminal::_AdjustCursorPosition(const COORD proposedPosition)
...
+        {
+            // scroll if no selection is active
+            scrollToOutput = scrollToOutput && !IsSelectionActive();
+        }
+        if (WI_IsFlagSet(_snapOnOutput, static_cast<int>(SnapOnOutputFlag::AtBottom)))
+        {
+            scrollToOutput = scrollToOutput && viewportAtBottom;
+        }
+
+        // use scrollToOutput to enforce SnapOnInput
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/Terminal.cpp:832]:
```suggestion
        // use scrollToOutput to enforce SnapOnOutput
```

**Comment by johnburnett** [comment]
This is probably coming in way to late, but having just found this - can we call this "scrollOnOutput"? The only reason I found this PR is because it's linked through other issues that reference "scroll" in their name. "snap" feels very obscure compared to what this setting actually controls (the scroll position).

**Comment by carlos-zamora** [comment]
> This is probably coming in way to late, but having just found this - can we call this "scrollOnOutput"? The only reason I found this PR is because it's linked through other issues that reference "scroll" in their name. "snap" feels very obscure compared to what this setting actually controls (the scroll position).

Fair request. Copying this comment over to #2529 for discussion there.

**Comment by carlos-zamora** [comment]
> Add a test?

I think we don't have scrolling-related tests, but I'll double check. So nobody come along and automerge this plz!

**Comment by carlos-zamora** [comment]
> block for a test

I hate that you're in the right for doing this... hahaha

**Comment by carlos-zamora** [comment]
Ok. I spent most of today trying to add a test to `ScrollTest`. I copied it below.

That should totally work. All I'm doing is writing lines of text to the buffer, and comparing the scroll offset to what it should be.

It doesn't though because we need a callback to the scrollback in TermControl. Spoke with Dustin and this just doesn't feel right. Like, scrolling should be pretty self sustained.

So, I created #6842 to track this better.

<details> <summary> Test Code </summary>

```c++

void ScrollTest::TestConditionalAutoscrolling()
{
    // When new output is generated, we should only scroll to it when...
    // - no selection is active, and
    // - the viewport is already at the bottom of the scroll history

    Log::Comment(L"Watch out - this test takes a while to run, and won't "
                 L"output anything unless in encounters an error. This is expected.");

    auto& termTb = *_term->_buffer;

    const auto totalBufferSize = termTb.GetSize().Height();

    // We're outputting like 18000 lines of text here, so emitting 18000*4 lines
    // of output to the console is actually quite unnecessary
    WEX::TestExecution::SetVerifyOutput settings(WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);

    // Emit a bunch of newlines to test scrolling.
    auto scrollAmount = 0;
    for (auto currentRow = 0; currentRow < totalBufferSize * 2; currentRow++)
    {
        *_scrollBarNotification = std::nullopt;
        _renderTarget->Reset();

        _term->_WriteBuffer(L"X\r\n");

        if (currentRow < TerminalViewHeight)
        {
            // no scrolling is expected
            VERIFY_ARE_EQUAL(0, _term->_scrollOffset);
        }
        else if (currentRow < TerminalViewHeight * 2)
        {
            // keep track of scrolling
            ++scrollAmount;
            VERIFY_ARE_EQUAL(scrollAmount, _term->_scrollOffset);
        }
        else if (currentRow == TerminalViewHeight * 2)
        {
            // create a selection
            Log::Comment(L"Create a selection to pause scrolling");
            _term->SelectNewRegion({ 0, static_cast<SHORT>(currentRow) }, { 5, static_cast<SHORT>(currentRow) });

            VERIFY_ARE_EQUAL(scrollAmount, _term->_scrollOffset);
        }
        else if (currentRow < totalBufferSize)
        {
            // scrolling is unchanged because of selection
            VERIFY_ARE_EQUAL(scrollAmount, _term->_scrollOffset);
        }
        else if (scrollAmount > 0)
        {
            // circling the buffer
            // we should be scrolling up to keep the same content in the viewport
            --scrollAmount;
            VERIFY_ARE_EQUAL(scrollAmount, _term->_scrollOffset);
        }
        else
        {
            // still circling the buffer
            // we can't scroll up anymore
            VERIFY_ARE_EQUAL(0, _term->_scrollOffset);
        }
    }
}
```

</details>

**Comment by brandon-kohn** [comment]
I'm using 1.17.1023 and when I'm compiling and scroll back to see an error after a few moments it jumps back down to the bottom of buffer. This happens whether I select some text or not.

**Pull request merged**